src/java/main/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -99,11 +100,15 @@ public void close() {
                         + Long.toHexString(sessionId));
             }
 
-            synchronized (factory.ipMap) {
-                Set<NettyServerCnxn> s =
-                    factory.ipMap.get(((InetSocketAddress)channel
-                            .getRemoteAddress()).getAddress());
-                s.remove(this);
+            InetAddress address = ((InetSocketAddress) channel.getRemoteAddress()).getAddress();
+            Set<NettyServerCnxn> s = factory.ipMap.get(address);
```

Comment:
This management of the `ipMap` looks like something that should be done by the factory, not the connection. should we move this logic to the factory?